pkg/driver/backup_gc.go

```diff
@@ -0,0 +1,336 @@
+	klog.Info("ZFS backup informer cache synced successfully")
+	return nil
+}
+
+// handleBackupCreation processes new ZFSBackup resources when they are created.
+// It validates whether the previous snapshot references are valid and initiates cleanup
+// of invalid references to maintain backup chain integrity.
+func (bgc *BackupGarbageCollector) handleBackupCreation(obj interface{}) {
+	backup, ok := obj.(*zfsapi.ZFSBackup)
+	if !ok {
+		klog.Errorf("Expected ZFSBackup, got %T", obj)
+		return
+	}
+
+	klog.InfoS("Processing backup creation", "backupName", backup.Name)
+
+	// Check if the backup references a prevSnapName that doesn't exist
+	if backup.Spec.PrevSnapName != "" {
+		go bgc.validateBackupPrevSnapName(backup)
+	}
+}
+
+// handleBackupUpdate processes changes to existing ZFSBackup resources.
+// It specifically handles changes to the previous snapshot reference to ensure
+// the backup chain remains valid after updates.
+func (bgc *BackupGarbageCollector) handleBackupUpdate(oldObj, newObj interface{}) {
+	oldBackup, ok := oldObj.(*zfsapi.ZFSBackup)
+	if !ok {
+		klog.Errorf("Expected ZFSBackup, got %T", oldObj)
+		return
+	}
+
+	newBackup, ok := newObj.(*zfsapi.ZFSBackup)
+	if !ok {
+		klog.Errorf("Expected ZFSBackup, got %T", newObj)
+		return
+	}
+
+	klog.InfoS("Processing backup update",
+		"backupName", newBackup.Name,
+		"prevSnapshot", newBackup.Spec.PrevSnapName)
+
+	// If prevSnapName was added or changed, validate it
+	if oldBackup.Spec.PrevSnapName != newBackup.Spec.PrevSnapName && newBackup.Spec.PrevSnapName != "" {
+		go bgc.validateBackupPrevSnapName(newBackup)
+	}
+}
+
+// handleBackupDeletion processes ZFSBackup deletion events and cleans up any
+// child backups that reference the deleted backup as their previous snapshot.
+// This ensures the backup chain integrity by removing dependent backups when
+// a parent backup is removed.
+func (bgc *BackupGarbageCollector) handleBackupDeletion(obj interface{}) {
+	backup, ok := obj.(*zfsapi.ZFSBackup)
+	if !ok {
+		// In case of delete event, we might get a DeletedFinalStateUnknown instead of the object
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Couldn't get object from tombstone %#v", obj)
+			return
+		}
+		backup, ok = tombstone.Obj.(*zfsapi.ZFSBackup)
+		if !ok {
+			klog.Errorf("Tombstone contained object that is not a ZFSBackup: %#v", obj)
+			return
+		}
+	}
+
+	klog.InfoS("Processing backup deletion",
+		"backupName", backup.Name,
+		"volumeName", backup.Spec.VolumeName,
+		"snapshot", backup.Spec.SnapName)
+
+	// When a backup is deleted, find and delete any backups that reference it as prevSnapName
+	go bgc.deleteBackupsReferencingDeletedBackup(backup)
+}
+
+// validateBackupPrevSnapName checks if the previous snapshot referenced by a backup exists.
+// If the reference doesn't exist, the backup is deleted to maintain chain integrity.
+func (bgc *BackupGarbageCollector) validateBackupPrevSnapName(backup *zfsapi.ZFSBackup) {
```

Comment:
Was a bit confused at first, is this name more understandable?
```suggestion
func (bgc *BackupGarbageCollector) deleteOrphanedIncBackup(backup *zfsapi.ZFSBackup) {
```

Reply:
Yes, I like it more too. Fixed on a7aa1d0a0b76d835f94edea295bee65bac0da5b1